plugins/Toolbox/src/Toolbox.py

```diff
@@ -0,0 +1,515 @@
+        local_package = self._package_manager.getInstalledPackageInfo(package_id)
+        if local_package is None:
+            return False
+
+        remote_package = None
+        for package in self._metadata["packages"]:
+            if package["package_id"] == package_id:
+                remote_package = package
+        if remote_package is None:
+            return False
+
+        local_version = local_package["package_version"]
+        remote_version = remote_package["package_version"]
+        return Version(remote_version) > Version(local_version)
+
+    @pyqtSlot(str, result = bool)
+    def isInstalled(self, package_id: str) -> bool:
+        return self._package_manager.isPackageInstalled(package_id)
+
+    @pyqtSlot(str, result = bool)
+    def isEnabled(self, package_id: str) -> bool:
+        if package_id in self._plugin_registry.getActivePlugins():
+            return True
+        return False
+
+    def loadingComplete(self) -> bool:
+        populated = 0
+        for list in self._metadata.items():
+            if len(list) > 0:
+                populated += 1
+        if populated == len(self._metadata.items()):
+            return True
+        return False
+
+
+
+    # Make API Calls
+    # --------------------------------------------------------------------------
+    def _makeRequestByType(self, type: str):
+        Logger.log("i", "Toolbox: Requesting %s metadata from server.", type)
+        request = QNetworkRequest(self._request_urls[type])
+        request.setRawHeader(*self._request_header)
+        self._network_manager.get(request)
+
+    @pyqtSlot(str)
+    def startDownload(self, url: str):
+        Logger.log("i", "Toolbox: Attempting to download & install package from %s.", url)
+        url = QUrl(url)
+        self._download_request = QNetworkRequest(url)
+        self._download_request.setAttribute(QNetworkRequest.RedirectPolicyAttribute, QNetworkRequest.NoLessSafeRedirectPolicy)
+        self._download_request.setRawHeader(*self._request_header)
+        self._download_reply = self._network_manager.get(self._download_request)
+        self.setDownloadProgress(0)
+        self.setIsDownloading(True)
+        self._download_reply.downloadProgress.connect(self._onDownloadProgress)
+
+    @pyqtSlot()
+    def cancelDownload(self):
+        Logger.log("i", "Toolbox: User cancelled the download of a plugin.")
+        self.resetDownload()
+        return
+
+    def resetDownload(self):
+        self._download_reply.abort()
+        self._download_reply.downloadProgress.disconnect(self._onDownloadProgress)
+        self._download_reply = None
+        self._download_request = None
+        self.setDownloadProgress(0)
+        self.setIsDownloading(False)
+
+
+
+    # Handlers for Network Events
+    # --------------------------------------------------------------------------
+    def _onNetworkAccessibleChanged(self, accessible: int):
+        if accessible == 0:
+            self.resetDownload()
```

Comment:
What does the value of `0` mean? Could you document this function?

Reply:
No idea. :/ That's something that wasn't changed by me. I think @sedwards2009 or @nallath wrote that a long time ago.

I could investigate in the network manager docs but that would have been pretty low on my priorities.